.github/workflows/python_tests.yml

```diff
@@ -9,7 +9,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.8, 3.9]
+        python-version: [3.8.9, 3.9.4]
```

Comment:
I am not a fan of pinning micro versions in CI.  I would rather that GitHub Actions tells us that a new release breaks stuff rather than users telling us.  How much time is using venv inside a venv really saving us?  Is it enough test time to offset the complexity that we are adding?

Reply:
The time difference is ~20 seconds ([non-venv-cached](https://github.com/internetarchive/openlibrary/runs/2355410178#step:5:1) | [venv-cached](https://github.com/internetarchive/openlibrary/pull/5057/checks?check_run_id=2358887923#step:6:1))

Thinking more about it, it does seem to be worth the complexity added. Let's see what @cdrini thinks about this.

Reply:
I tried re-running them both again to compare them, and times were:

| V | Time with ~/.cache | Time with venv |
| -- | -- | -- |
| 3.8* | 72s | 39s | 
| 3.9* | 52s | 37s |

So it does seem faster... 

Hmmm; question, where does pip install its stuff usually when there isn't a venv? Could we cache that instead?

Reply:
`python3 -m pip --version ` # will display the path to site-packages

Reply:
I don't think only caching the site-packages is the solution as the CLI tools such as `black`, `mypy`, etc. has their entry points in the `bin` directory which is also being cached when we use the `venv` method.

Reply:
I think I found a solution! This article here proposes caching the entire python env, and it seems to be resilient enough to let us specify eg `3.9` in the config! Can we try the second approach here? https://medium.com/ai2-blog/python-caching-in-github-actions-e9452698e98d

Reply:
Awesome! Nice find. It works like a charm: 
[2m 44s](https://github.com/internetarchive/openlibrary/runs/2425914516) -> [1m 24s](https://github.com/internetarchive/openlibrary/runs/2425935598)

It took some time to save the cache as it is around 125MB in size, but that will only be when we upgrade any of our dependencies or Python version.